### PR TITLE
JS Driver: add `concurrency` optarg to `eachAsync`

### LINKED ISSUE #5509: Add optarg to `eachAsync` for managing concurrency
Now that we have `eachAsync` for automatically awaiting async operations before continuing to read from the cursor, it would be nice if the concurrency could be controlled. This would essentially be equivalent to [`Promise.map(..., { concurrency: <int> })`](http://bluebirdjs.com/docs/api/promise.map.html#map-option-concurrency) from bluebird.

While this can be done in user-code with relative ease, I think it's useful enough (namely in tuning the performance of bulk operations) to consider implementing.

```js
const concurrency = 10;
const pending = [];

cursor.eachAsync((job) => {
  Promise.resolve().then(() => {
    if (pending.length >= concurrency) {
      return Promise.any(pending);
    }
  }).then(() => {
    let p = asyncHandler(job).then(() =>
      pending.splice(pending.indexOf(p), 1);
    );
    pending.push(p);
  });
});
```

Proposal:

```js
cursor.eachAsync(asyncHandler, { concurrency: 10 });
```

### PR DESCRIPTION
**Pull request by marshall007**
Fixes #5509. Note: this is on top of my changes for #5478 since they would conflict.

### CODE REVIEW

**Comment by danielmewes** [comment]
Thanks @marshall007 !

Assigning to @nighelles for review.

File: drivers/javascript/cursor.coffee [hunk 1]
```diff
@@ -211,50 +209,65 @@ class IterableResult
...
-                @_next nextCb
-            else
-                onFinished?()
-        @_next nextCb
+        nextCb = =>
+            @_next().then (data) ->
+                return nextCb() if cb(null, data) isnt false
+            .catch (err) ->
+                return if err?.message is 'No more rows in the cursor.'
+
```
Comment (nighelles) [inline drivers/javascript/cursor.coffee]: Do we lose the cb(err) call here?

File: drivers/javascript/cursor.coffee [hunk 2]
```diff
@@ -211,50 +209,65 @@ class IterableResult
...
             if @_closeCbPromise?
                 return Promise.reject(new error.ReqlDriverError("Cursor is closed."))
             else
-                @_next().then (data) ->
-                    return cb(data) if cb.length <= 1 # either synchronous or awaits promise
-                    return Promise.fromNode (handler) -> cb(data, handler) # callback-style async
-                .then (result) ->
-                    return nextCb()
+                return @_next().then (data) ->
+                    return data if pending.length < options.concurrency
```
Comment (nighelles) [inline drivers/javascript/cursor.coffee]: If options is undefined, is this always false?

Comment (marshall007) [inline drivers/javascript/cursor.coffee]: `options.concurrency` will always either be what the user specified or `1` by the time we make it to this line.

Comment (nighelles) [inline drivers/javascript/cursor.coffee]: Ok, missed the default.

File: drivers/javascript/cursor.coffee (hunk 1 above)
Comment (marshall007) [inline drivers/javascript/cursor.coffee]: Whoops, yea good catch.

**Comment by larkost** [comment]
@marshall007: I have (finally) pushed my branch with an overhaul of the JavaScript connection tests in branch `larkost/5413-javascript-connection-tests`. A lot of this is about better defining our behavior, but there are a lot of failures, both with our current version of the driver, as well as your revision (largely the same issues). I know that @nighelles is looking into these areas, but it seems we are all working on intersecting changes.

**Comment by nighelles** [comment]
Ok, now that these tests are up, I'll see about testing everything with theses new changes.

**Comment by marshall007** [comment]
@larkost nice, this is great! I'll hold off on making any more changes so we don't step on each other's toes trying to get the tests to pass.

**Comment by nighelles** [comment]
This was merged into next along with some other eachAsync fixes.

**Pull request closed without merging**